torch/__init__.py

```diff
@@ -193,6 +194,42 @@ def _load_global_deps():
 if False:
     import torch._C as _C
 
+# Check to see if we can load C extensions, and if not provide some guidance
+# on what the problem might be.
+import torch._C as _C_for_compiled_check
+if _C_for_compiled_check.__file__ is None:
```

Comment:
@ShawnZhong Can you point where it is documented what the meaning of `__file__` on C extensions is supposed to mean?

Reply:
To clarify, I don't want to be relying on undocumented "accidents" in Python, for something that is supposed to help users debug. If we're trying to discover user misconfiguration and report against it, the way we test for it needs to be 100% correct; otherwise, we're just making the problem worse when the detection mechanism is wrong.

Reply:
From https://docs.python.org/3/reference/datamodel.html#the-standard-type-hierarchy: 

> __file__ is the pathname of the file from which the module was loaded, if it was loaded from a file. The __file__ attribute may be missing for certain types of modules, such as C modules that are statically linked into the interpreter; for extension modules loaded dynamically from a shared library, it is the pathname of the shared library file.


So if `torch._C` is loaded from a shared library (`torch/_C.cpython-37m-x86_64-linux-gnu.so` in my case), the `__file__` attribute is the pathname of that file. If the file is missing, then `torch._C` is loaded from the stub folder `torch/_C`, and gives `None` for its `__file__` attribute. 

I am open to alternative methods to check if a module is loaded from C extensions. I have tried `ctypes.util.find_library` and `ctypes.cdll.LoadLibrary` but they didn't seem to work.

Reply:
I don't think stub file should ever be loaded; it's a pyi file and I'm pretty sure those get ignored....

Reply:
No, they shouldn't. I am saying that `torch._C` is loaded from the `torch/_C` in the absence of the actual shared library file, so there is no `ImportError` thrown when calling `from torch._C import *` in `__init__.py` in that case.

Reply:
Empirically the None check does work. (That's not proof that it's general, of course. But it is not ignored.) And I agree that absent that `torch._C` will resolve to `torch/_C`.

Reply:
I don't think so. There is no `torch/_C/__init__.py` file, so you should never resolve to this directory.

Reply:
You don't need an `__init__` file to resolve an import:
```
$ mkdir test123
$ python3
>>> import test123
>>> print(test123.__file__)
None
```

Reply:
Oh, this is a new Python 3.3 thing for implicit namespace packages 🤦